### PR TITLE
[mariadb-galera] fix persistent volume permissions

### PR DESCRIPTION
**Pull request by bashar-alkhateeb**
our Backplane  NFS driver does not support the change of fsGroup in case of root mismatch that's why we need to change the owner of the root directory of the volumes if we want to use persistent volumes with none root containers.
add privileged init container with all persistent volume mounted on it, which runs chown -R over all mounts.

Change-Id: I8c0d7dd94d7d24f4cd2ef1a7e2af10ed11d6139d

### CODE REVIEW

File: common/mariadb-galera/helm/templates/statefulset-mariadb.yaml [hunk 1]
```diff
@@ -115,6 +115,31 @@ spec:
           privileged: {{ $.Values.initContainers.cleanoscache.securityContext.privileged }}
           runAsUser: {{ $.Values.initContainers.cleanoscache.securityContext.runAsUser | default 0 | int }}
       {{- end }}
+      - name: fix-mariadb-fs-permissions
```
Comment (businessbean) [inline common/mariadb-galera/helm/templates/statefulset-mariadb.yaml:118]: Please make that initContainer optional like we have done it for the [clean-os-cache](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/templates/statefulset-mariadb.yaml#L106) container. Please also create a [mariab.fixFsPermissions](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L66) boolean option (default false) and add some documentation similar to [wipeDataAndLog](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L69-L71)

File: common/mariadb-galera/helm/templates/statefulset-proxysql.yaml [hunk 2]
```diff
@@ -107,6 +107,31 @@ spec:
         securityContext:
           privileged: {{ $.Values.initContainers.increaseMapCount.securityContext.privileged | default true }}
           runAsUser: {{ $.Values.initContainers.increaseMapCount.securityContext.runAsUser | default 0 | int }}
+      - name: fix-proxysql-fs-permissions
```
Comment (businessbean) [inline common/mariadb-galera/helm/templates/statefulset-proxysql.yaml:110]: Please make that initContainer optional like we have done it for the [clean-os-cache](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/templates/statefulset-mariadb.yaml#L106) container. Please also create a [proxy.proxysql.fixFsPermissions](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L693) boolean option (default false) and add some documentation similar to [wipeDataAndLog](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L69-L71)

File: common/mariadb-galera/helm/templates/statefulset-proxysql.yaml [hunk 3]
```diff
@@ -107,6 +107,31 @@ spec:
...
+        image: "{{ $.Values.image.os.registry }}/{{ $.Values.image.os.project }}/{{ $.Values.image.os.databasename }}:{{ $.Values.image.os.databaseversion }}-{{ $.Values.image.os.imageversion | int }}"
+        imagePullPolicy: {{ $.Values.image.os.pullPolicy | default "IfNotPresent" | quote }}
+        command:
+        - sh
+        - -c
+        - 'chown -R {{ $.Values.userId.proxy | default 3100 | int }}:{{ $.Values.groupId.proxy | default 3100 | int }} /mnt/*'
+        securityContext:
+          runAsUser: {{ $.Values.initContainers.fixProxysqlFsPermissions.securityContext.runAsUser | default 0 | int }}
```
Comment (businessbean) [inline common/mariadb-galera/helm/templates/statefulset-proxysql.yaml:118]: Please add a `initContainers.fixProxysqlFsPermissions.securityContext.runAsUser` option to the values.yaml similar to [cleanoscache](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L856) and add some documentation similar to [wipeDataAndLog](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L69-L71)

File: common/mariadb-galera/helm/templates/statefulset-mariadb.yaml [hunk 4]
```diff
@@ -115,6 +115,31 @@ spec:
...
+        image: "{{ $.Values.image.os.registry }}/{{ $.Values.image.os.project }}/{{ $.Values.image.os.databasename }}:{{ $.Values.image.os.databaseversion }}-{{ $.Values.image.os.imageversion | int }}"
+        imagePullPolicy: {{ $.Values.image.os.pullPolicy | default "IfNotPresent" | quote }}
+        command:
+        - sh
+        - -c
+        - 'chown -R {{ $.Values.userId.database | default 101 | int }}:{{ $.Values.groupId.database | default 101 | int }} /mnt/*'
+        securityContext:
+          runAsUser: {{ $.Values.initContainers.fixFsPermissions.securityContext.runAsUser | default 0 | int }}
```
Comment (businessbean) [inline common/mariadb-galera/helm/templates/statefulset-mariadb.yaml:126]: Please rename `initContainers.fixFsPermissions...` to `initContainers.fixMariaDBFsPermissions...` and please add a `initContainers.fixMariaDBFsPermissions.securityContext.runAsUser` option to the values.yaml similar to [cleanoscache](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L856) and add some documentation similar to [wipeDataAndLog](https://github.com/sapcc/helm-charts/blob/8188e613e78536078c381ad3f4b04daa9c55494b/common/mariadb-galera/helm/values.yaml#L69-L71)

**Review by businessbean — CHANGES_REQUESTED** [review]
After implementing the described changes also do these steps:
- update the [CHANGELOG](https://github.com/sapcc/helm-charts/blob/master/common/mariadb-galera/CHANGELOG.md)
- `helm unittest helm` and update the rules if required
- `helm-docs --chart-search-root=helm --output-file=../README.md`
- `helm package helm`
- `helm push mariadb-galera-0.27.1.tgz oci://keppel.eu-de-1.cloud.sap/ccloud-helm/`

Later we can further automate these steps...

**Comment by bashar-alkhateeb** [comment]
> After implementing the described changes also do these steps:
> 
> * update the [CHANGELOG](https://github.com/sapcc/helm-charts/blob/master/common/mariadb-galera/CHANGELOG.md)
> * `helm unittest helm` and update the rules if required
> * `helm-docs --chart-search-root=helm --output-file=../README.md`
> * `helm package helm`
> * `helm push mariadb-galera-0.27.1.tgz oci://keppel.eu-de-1.cloud.sap/ccloud-helm/`
> 
> Later we can further automate these steps...

only related unittests have been fixed

File: common/mariadb-galera/helm/templates/statefulset-mariadb.yaml (hunk 1 above)
Comment (bashar-alkhateeb) [reply common/mariadb-galera/helm/templates/statefulset-mariadb.yaml:119]: done!

File: common/mariadb-galera/helm/templates/statefulset-proxysql.yaml (hunk 3 above)
Comment (bashar-alkhateeb) [reply common/mariadb-galera/helm/templates/statefulset-proxysql.yaml:119]: done!

File: common/mariadb-galera/helm/templates/statefulset-mariadb.yaml (hunk 4 above)
Comment (bashar-alkhateeb) [reply common/mariadb-galera/helm/templates/statefulset-mariadb.yaml]: done!

File: common/mariadb-galera/README.md [hunk 5]
```diff
@@ -312,6 +312,8 @@ docker build --build-arg BASE_SOFT_NAME=ubuntu --build-arg BASE_SOFT_VERSION=22.
 | image.pullSecrets.secretname.registry | string | `nil` | the hostname of the container registry that should be used for the pull secret |
 | initContainers.cleanoscache.securityContext.privileged | bool | `true` | required to configure `/proc/sys/vm/drop_caches` in the init phase |
 | initContainers.cleanoscache.securityContext.runAsUser | int | 0 | required to configure `/proc/sys/vm/drop_caches` in the init phase |
+| initContainers.fixMariaDBFsPermissions.securityContext.privileged | bool | `true` | required to configure `/proc/sys/vm/drop_caches` in the init phase |
```
Comment (businessbean) [inline common/mariadb-galera/README.md:315]: copy/paste error...
```
required to configure `/proc/sys/vm/drop_caches` in the init phase
```

File: common/mariadb-galera/README.md [hunk 6]
```diff
@@ -312,6 +312,8 @@ docker build --build-arg BASE_SOFT_NAME=ubuntu --build-arg BASE_SOFT_VERSION=22.
...
+| initContainers.fixMariaDBFsPermissions.securityContext.runAsUser | int | 0 | required to configure `/proc/sys/vm/drop_caches` in the init phase |
```
Comment (businessbean) [inline common/mariadb-galera/README.md:316]: copy/paste error...
```
required to configure `/proc/sys/vm/drop_caches` in the init phase
```

File: common/mariadb-galera/helm/templates/statefulset-proxysql.yaml [hunk 7]
```diff
@@ -107,6 +107,33 @@ spec:
...
+      {{- if $.Values.proxy.proxysql.fixFsPermissions}}
+      - name: fix-proxysql-fs-permissions
+        image: "{{ $.Values.image.os.regiostry }}/{{ $.Values.image.os.project }}/{{ $.Values.image.os.databasename }}:{{ $.Values.image.os.databaseversion }}-{{ $.Values.image.os.imageversion | int }}"
+        imagePullPolicy: {{ $.Values.image.os.pullPolicy | default "IfNotPresent" | quote }}
+        command:
+        - sh
+        - -c
+        - 'chown -R {{ $.Values.userId.proxy | default 3100 | int }}:{{ $.Values.groupId.proxy | default 3100 | int }} /mnt/*'
+        securityContext:
+          runAsUser: {{ $.Values.initContainers.fixProxysqlFsPermissions.securityContext.runAsUser | default 0 | int }}
```
Comment (businessbean) [inline common/mariadb-galera/helm/templates/statefulset-proxysql.yaml:119]: `initContainers.fixProxysqlFsPermissions...` is missing in the `values.yaml`

**Review by businessbean — CHANGES_REQUESTED** [review]
some minor fixes are required

File: common/mariadb-galera/README.md [hunk 8]
```diff
@@ -312,6 +312,10 @@ docker build --build-arg BASE_SOFT_NAME=ubuntu --build-arg BASE_SOFT_VERSION=22.
 | image.pullSecrets.secretname.registry | string | `nil` | the hostname of the container registry that should be used for the pull secret |
 | initContainers.cleanoscache.securityContext.privileged | bool | `true` | required to configure `/proc/sys/vm/drop_caches` in the init phase |
 | initContainers.cleanoscache.securityContext.runAsUser | int | 0 | required to configure `/proc/sys/vm/drop_caches` in the init phase |
+| initContainers.fixMariaDBFsPermissions.securityContext.privileged | bool | `true` | required to fix FS permissions of mariadb persistent volumes in the init phase |
+| initContainers.fixMariaDBFsPermissions.securityContext.runAsUser | int | 0 | required to fix FS permissions of proxysql persistent volumes in the init phase |
+| initContainers.fixProxysqlFsPermissions.securityContext.privileged | bool | `true` | required to configure `/proc/sys/vm/drop_caches` in the init phase |
```
Comment (businessbean) [inline common/mariadb-galera/README.md:317]: another copy/paste typo

File: common/mariadb-galera/README.md [hunk 9]
```diff
@@ -312,6 +312,10 @@ docker build --build-arg BASE_SOFT_NAME=ubuntu --build-arg BASE_SOFT_VERSION=22.
...
+| initContainers.fixProxysqlFsPermissions.securityContext.runAsUser | int | 0 | required to configure `/proc/sys/vm/drop_caches` in the init phase |
```
Comment (businessbean) [inline common/mariadb-galera/README.md:318]: another copy/paste typo

**Review by businessbean — CHANGES_REQUESTED** [review]
two small typos remaining

**Review by businessbean — APPROVED** [review]
lgtm

**Comment by businessbean** [comment]
> @businessbean can we maintain code owners for this helm-chart to avoid pulling in the repo admins as reviewers? Thanks

Yes, but their is already some CODEOWNERS maintenance ongoing. Let me check the current status of the PRs...

**Comment by businessbean** [comment]
> @businessbean can we maintain code owners for this helm-chart to avoid pulling in the repo admins as reviewers? Thanks

done

**Comment by Carthaca** [comment]
When I added this for rabbitmq, Fabian gave me a hint how to make this less expensive
https://github.com/sapcc/helm-charts/pull/6133#issuecomment-1985661106

You might want to add this here, too

**Pull request merged**